### ISSUE TITLE
Recent addition of vterm breaks Emacs startup on macOS

### ISSUE THREAD

**Issue by mpereira** [magit/with-editor#85]
[This commit](https://github.com/magit/with-editor/commit/a339d6590096829a274c1901e407873e1fc1016e) breaks Emacs startup on macOS because vterm is currently [failing to compile on macOS](https://github.com/akermu/emacs-libvterm/issues/433).

Reproducing the issue should be possible by using latest MELPA package versions of with-editor and vterm, removing all byte-compiled files, and starting Emacs.

Emacs backtrace:
```
Debugger entered--Lisp error: (error "Vterm will not work until ‘vterm-module’ is compil...")
  signal(error ("Vterm will not work until ‘vterm-module’ is compil..."))
  error("Vterm will not work until `vterm-module' is compil...")
  (if (or vterm-always-compile-module (y-or-n-p "Vterm needs `vterm-module' to work.  Compile it no...")) (progn (vterm-module-compile) (require 'vterm-module)) (error "Vterm will not work until `vterm-module' is compil..."))
  (if (require 'vterm-module nil t) nil (if (or vterm-always-compile-module (y-or-n-p "Vterm needs `vterm-module' to work.  Compile it no...")) (progn (vterm-module-compile) (require 'vterm-module)) (error "Vterm will not work until `vterm-module' is compil...")))
  eval-buffer(#<buffer  *load*-823129> nil "/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/..." nil t)  ; Reading at buffer position 4829
  load-with-code-conversion("/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/..." "/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/..." t t)
  #<subr require>(vterm nil t)
  ad-Advice-require(#<subr require> vterm nil t)
  apply(ad-Advice-require #<subr require> (vterm nil t))
  require(vterm nil t)
  (progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (require 'vterm nil t) (require 'warnings nil t))
  (progn (progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (require 'vterm nil t) (require 'warnings nil t)))
  eval((progn (progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (require 'vterm nil t) (require 'warnings nil t))) t)
  #f(compiled-function (&rest body) "Like `progn', but evaluates the body at compile time if you're compiling.\nThus, the result of the body appears to the compiler as a quoted\nconstant.  In interpreted code, this is entirely equivalent to\n`progn', except that the value of the expression may be (but is\nnot necessarily) computed at load time if eager macro expansion\nis enabled." #<bytecode 0x1533d2cd9a2804>)((progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (require 'vterm nil t) (require 'warnings nil t)))
  (eval-when-compile (progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (require 'vterm nil t) (require 'warnings nil t)))
  eval-buffer(#<buffer  *load*-517981> nil "/Users/mpereira/.emacs.d/elpa/with-editor-20201026..." nil t)  ; Reading at buffer position 4212
  load-with-code-conversion("/Users/mpereira/.emacs.d/elpa/with-editor-20201026..." "/Users/mpereira/.emacs.d/elpa/with-editor-20201026.2027/with-editor.el" t t)
  #<subr require>(with-editor nil t)
  ad-Advice-require(#<subr require> with-editor nil t)
  apply(ad-Advice-require #<subr require> (with-editor nil t))
  require(with-editor nil t)
  (not (require 'with-editor nil t))
  (if (not (require 'with-editor nil t)) (display-warning 'use-package (format "Cannot load %s" 'with-editor) :error) (condition-case err (progn (add-hook 'eshell-mode-hook 'with-editor-export-editor) (add-hook 'term-exec-hook 'with-editor-export-editor) (add-hook 'shell-mode-hook 'with-editor-export-editor) (add-hook 'with-editor-mode-hook 'evil-insert-state) t) ((debug error) (funcall use-package--warning68 :config err))))
  (condition-case err (if (not (require 'with-editor nil t)) (display-warning 'use-package (format "Cannot load %s" 'with-editor) :error) (condition-case err (progn (add-hook 'eshell-mode-hook 'with-editor-export-editor) (add-hook 'term-exec-hook 'with-editor-export-editor) (add-hook 'shell-mode-hook 'with-editor-export-editor) (add-hook 'with-editor-mode-hook 'evil-insert-state) t) ((debug error) (funcall use-package--warning68 :config err)))) ((debug error) (funcall use-package--warning68 :catch err)))
  eval-buffer(#<buffer  *load*-2065> nil "/Users/mpereira/.emacs.d/configuration.el" nil t)  ; Reading at buffer position 107766
  load-with-code-conversion("/Users/mpereira/.emacs.d/configuration.el" "/Users/mpereira/.emacs.d/configuration.el" nil nil)
  #<subr load>("/Users/mpereira/.emacs.d/configuration.el" nil nil t nil)
  ad-Advice-load(#<subr load> "/Users/mpereira/.emacs.d/configuration.el" nil nil t)
  apply(ad-Advice-load #<subr load> ("/Users/mpereira/.emacs.d/configuration.el" nil nil t))
  load("/Users/mpereira/.emacs.d/configuration.el" nil nil t)
  load-file("/Users/mpereira/.emacs.d/configuration.el")
```

vterm compilation failure:
```
-- System libvterm not found: libvterm will be downloaded and compiled as part of the build process
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/build
[  8%] Performing update step for 'libvterm'
[ 16%] No patch step for 'libvterm'
[ 25%] No configure step for 'libvterm'
[ 33%] Performing build step for 'libvterm'
CC bin/vterm-ctrl.c
bin/vterm-ctrl.c:117:10: error: implicit declaration of function 'strdup' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  return strdup(csi);
         ^
bin/vterm-ctrl.c:117:10: note: did you mean 'strcmp'?
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/string.h:77:6: note: 'strcmp' declared here
int      strcmp(const char *__s1, const char *__s2);
         ^
bin/vterm-ctrl.c:117:10: warning: incompatible integer to pointer conversion returning 'int' from a function with result type 'char *' [-Wint-conversion]
  return strdup(csi);
         ^~~~~~~~~~~
bin/vterm-ctrl.c:142:10: error: implicit declaration of function 'strdup' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  return strdup(dcs);
         ^
bin/vterm-ctrl.c:142:10: warning: incompatible integer to pointer conversion returning 'int' from a function with result type 'char *' [-Wint-conversion]
  return strdup(dcs);
         ^~~~~~~~~~~
bin/vterm-ctrl.c:129:10: warning: result of comparison of constant 156 with expression of type 'char' is always false [-Wtautological-constant-out-of-range-compare]
    if(c == 0x9c) // ST
       ~ ^  ~~~~
3 warnings and 2 errors generated.
make[3]: *** [bin/vterm-ctrl] Error 1
make[2]: *** [libvterm-prefix/src/libvterm-stamp/libvterm-build] Error 2
make[1]: *** [CMakeFiles/libvterm.dir/all] Error 2
make: *** [all] Error 2
/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057
```

Emacs: 28.0.50
macOS: 10.15.5
vterm-20201004.2057
with-editor-20201026.202

**Comment by tarsius** [comment]
> Reproducing the issue should be possible by using latest MELPA package versions of with-editor and vterm, removing all byte-compiled files, and starting Emacs.

That and buying a mac.

Does this do the trick?

```elisp
(with-demoted-errors "Error: %S" (require 'vterm nil t))
```

**Comment by mpereira** [comment]
> That and buying a mac.

🙂 
 
> Does this do the trick?
> 
> (with-demoted-errors "Error: %S"
>   (require 'vterm nil t))

Only when `debug-on-error` is explicitly set to `nil` (the default is `startup`).

Otherwise, I get this: 
```
Debugger entered--Lisp error: (file-missing "Cannot open load file" "No such file or directory" "vterm-module")
  #<subr require>(vterm-module nil nil)
  ad-Advice-require(#<subr require> vterm-module)
  apply(ad-Advice-require #<subr require> vterm-module)
  require(vterm-module)
  (progn (vterm-module-compile) (require 'vterm-module))
  (if (or vterm-always-compile-module (y-or-n-p "Vterm needs `vterm-module' to work.  Compile it no...")) (progn (vterm-module-compile) (require 'vterm-module)) (error "Vterm will not work until `vterm-module' is compil..."))
  (if (require 'vterm-module nil t) nil (if (or vterm-always-compile-module (y-or-n-p "Vterm needs `vterm-module' to work.  Compile it no...")) (progn (vterm-module-compile) (require 'vterm-module)) (error "Vterm will not work until `vterm-module' is compil...")))
  eval-buffer(#<buffer  *load*-432468> nil "/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/..." nil t)  ; Reading at buffer position 4829
  load-with-code-conversion("/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/..." "/Users/mpereira/.emacs.d/elpa/vterm-20201004.2057/vterm.el" t t)
  #<subr require>(vterm nil t)
  ad-Advice-require(#<subr require> vterm nil t)
  apply(ad-Advice-require #<subr require> (vterm nil t))
  require(vterm nil t)
  (condition-case err (require 'vterm nil t) ((debug error) (message "Error: %S" err) nil))
  (condition-case-unless-debug err (require 'vterm nil t) (error (message "Error: %S" err) nil))
  (with-demoted-errors "Error: %S" (require 'vterm nil t))
  (progn (require 'dired nil t) (require 'eshell nil t) (require 'term nil t) (with-demoted-errors "Error: %S" (require 'vterm nil t)) (require 'warnings nil t))
```

---

I appreciate you taking a look so quickly.

**Comment by tarsius** [comment]
What about

```elisp
(condition-case err
    (require 'vterm nil t)  
  (error (message "Error: %S" err)))
```

**Comment by mpereira** [comment]
That fixes it.

**Comment by tarsius** [comment]
Just pushed a fix for the fix: 202ea07838fb59002e292a541470492fc6690283.
